### DISCUSSION TITLE
Notification or delivery problem | Help me | Django channels

### DISCUSSION THREAD

**Discussion by QuvonchbekBobojonov** [community/community#77304] [Programming Help]
## cunsumers.py

```py

from rest_framework_simplejwt.tokens import AccessToken
from channels.generic.websocket import WebsocketConsumer
import json
from asgiref.sync import async_to_sync

from apps.users.models import CustomUser

from .models import Notification


class NotificationConsumer(WebsocketConsumer):
    def __init__(self, *args, **kwargs):
        super().__init__(args, kwargs)
        self.user = None
        self.room_name = 'notification'
        self.room_group_name = 'notification_group'

    def connect(self):
        query = str(self.scope['query_string'])
        token = query.replace('b\'token=', '').replace('\'', '')
        user, success = self.get_user(token)
        if success:
            async_to_sync(self.channel_layer.group_add)(self.room_name, self.room_group_name)
            self.accept()
            self.send(text_data=json.dumps({
                'message': "notification connected"
            }))
            for notification in Notification.objects.filter(user=user, is_seen=False).all():
                message = {
                    'user': self.user.full_name,
                    'notification_count': notification.get_is_seen_notification_count(),
                    'notification': notification.notification,
                    'notification_json': notification.notification_json,
                    'date_sent': notification.joined_at
                }
                async_to_sync(self.channel_layer.group_send)(
                    self.room_group_name,
                    {
                        'type': 'chat.message',
                        'message': message
                    }
                )
        else:
            print(user)
            self.close()

    def receive(self, text_data):
        print('receive', text_data)

    def disconnect(self, close_code):
        user_id = int(self.room_group_name.replace('user_', ''))
        user = CustomUser.objects.get(id=user_id)
        user.status = False
        user.save()
        self.send(text_data=json.dumps({
            'message': "notification disconnected"
        }))
        async_to_sync(self.channel_layer.group_discard)(self.room_name, self.room_group_name)
        self.close()

    def chat_message(self, event):
        message = event["message"]
        print(message)
        self.send(text_data=json.dumps({"message": message}))

    def get_user(self, token):
        try:
            access_token = AccessToken(token)
            user = access_token.payload.get('user_id')
            self.room_group_name = f'user_{user}'
            user_obj = CustomUser.objects.get(id=user)
            user_obj.status = True
            user_obj.save()
            self.user = user_obj
            return user_obj, True
        except Exception as e:
            return str(e), False


```

## models.py 

```py

from django.db import models
from django.utils.timezone import datetime
from jsonfield import JSONField
import channels.layers
from asgiref.sync import async_to_sync


class Notification(models.Model):
    user = models.ForeignKey('users.CustomUser', on_delete=models.SET_NULL, null=True)
    notification = models.TextField(null=True)
    notification_json = JSONField(blank=True, null=True)
    date_sent = models.DateTimeField(auto_now_add=True, blank=True)
    is_seen = models.BooleanField(default=False)
    joined_at = models.DateTimeField(auto_now_add=True, blank=True)

    def __str__(self):
        return f'{self.user.full_name} - {self.notification}'

    def save(self, *args, **kwargs):
        room_group_name = f'user_{self.user.id}'
        print(room_group_name)
        channels_layers = channels.layers.get_channel_layer()
        async_to_sync(channels_layers.group_send)(
            room_group_name,
            {
                'type': 'chat.message',
                'message': {
                    'user': self.user.full_name,
                    'notification_count': self.get_is_seen_notification_count(),
                    'notification': self.notification,
                    'notification_json': self.notification_json,
                    'date_sent': self.joined_at
                }
            }
        )
        return super().save(*args, **kwargs)

    @classmethod
    def get_is_seen_notification_count(cls):
        return cls.objects.filter(is_seen=False).count()


```

## settings.py

```py

CHANNEL_LAYERS = {
    "default": {
        "BACKEND": "channels_redis.core.RedisChannelLayer",
        "CONFIG": {
            "hosts": [('localhost', 6379)],
        }
    }
}

```

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by CheetahCodes21** [comment]
### Can you please specify your issue properly
## As of what I understood is:
The provided code defines a `NotificationConsumer` class that handles WebSocket connections for receiving notifications. It utilizes Django Channels and Redis to establish a real-time communication channel between the server and the client.

Here's a breakdown of the code:

**consumers.py**

The `NotificationConsumer` class inherits from `WebsocketConsumer` and handles the WebSocket connection lifecycle.

- `connect()`:
  - Extracts the token from the query string.
  - Validates the token and retrieves the associated user.
  - If the token is valid, joins the user to the notification group and sends a connection message.
  - If the token is invalid, logs the error and closes the connection.

- `receive()`:
  - Receives text data from the client and prints it to the console.

- `disconnect()`:
  - Updates the user's status to inactive.
  - Sends a disconnection message.
  - Removes the user from the notification group and closes the connection.

- `chat_message()`:
  - Receives a message from the server and sends it to the client.

- `get_user()`:
  - Validates the provided token and extracts the user ID.
  - Retrieves the corresponding `CustomUser` object and updates their status to active.
  - Sets the room group name based on the user ID and returns the user object.

**models.py**

The `Notification` model defines the structure of notifications.

- `user`: Foreign key to the associated `CustomUser` object.
- `notification`: Text field containing the notification message.
- `notification_json`: JSONField to store additional notification data.
- `date_sent`: Date and time when the notification was created.
- `is_seen`: Boolean flag indicating whether the notification has been seen by the user.
- `joined_at`: Date and time when the notification was sent to the WebSocket consumer.

- `__str__()`: Defines the string representation of the notification.

- `save()`:
  - Saves the notification model instance.
  - Constructs a notification message using the notification details and user information.
  - Sends the notification message to the user's notification group using `channels_layers`.

- `get_is_seen_notification_count()`: Class method that returns the total number of unread notifications.

**settings.py**

Defines the channel layer configuration for using Redis as the backend.

- `CHANNEL_LAYERS`: Sets the default channel layer to use Redis.
- `CONFIG`: Specifies the Redis host and port.

This setup enables the `NotificationConsumer` to handle real-time notification delivery to connected users via WebSocket connections.
